File: iosApp/iosApp/platformServices/IapServiceImpl.swift
//
//  IapServiceImpl.swift
//  iosApp
//
//  Created by Stanislav Aleshin on 26.06.2025.
//  Copyright © 2025 orgName. All rights reserved.
//

import Foundation
import shared

@objc(CommonIapServiceImpl)
public class IapServiceImpl: NSObject, CommonIapService {

    public func __confirmPurchase(purchaseId: String, developerPayload: String?, completionHandler: @escaping ((any Error)?) -> Void) {
        Task {
            do {
                try await __confirmPurchase(purchaseId: purchaseId, developerPayload: developerPayload)
                completionHandler(nil)
            } catch {
                completionHandler(error)
            }
        }
    }

    @nonobjc
    public func __confirmPurchase(purchaseId: String, developerPayload: String?) async throws {
        try await confirmPurchase(purchaseId: purchaseId, developerPayload: developerPayload)
    }

    public func __deletePurchase(purchaseId: String, completionHandler: @escaping ((any Error)?) -> Void) {
        Task {
            do {
                try await __deletePurchase(purchaseId: purchaseId)
                completionHandler(nil)
            } catch {
                completionHandler(error)
            }
        }
    }

    @nonobjc
    public func __deletePurchase(purchaseId: String) async throws {
        try await deletePurchase(purchaseId: purchaseId)
    }

    public func __fetchProducts(ids: [String], completionHandler: @escaping ([CommonIapProduct]?, (any Error)?) -> Void) {
        Task {
            do {
                let result = try await __fetchProducts(ids: ids)
                completionHandler(result, nil)
            } catch {
                completionHandler(nil, error)
            }
        }
    }

    @nonobjc
    public func __fetchProducts(ids: [String]) async throws -> [CommonIapProduct] {
        return try await fetchProducts(ids: ids)
    }

    public func __fetchPurchaseInfo(purchaseId: String, completionHandler: @escaping (CommonIapPurchase?, (any Error)?) -> Void) {
        Task {
            do {
                let result = try await __fetchPurchaseInfo(purchaseId: purchaseId)
                completionHandler(result, nil)
            } catch {
                completionHandler(nil, error)
            }
        }
    }

    @nonobjc
    public func __fetchPurchaseInfo(purchaseId: String) async throws -> CommonIapPurchase? {
        return try await fetchPurchaseInfo(purchaseId: purchaseId)
    }

    public func __fetchPurchases(completionHandler: @escaping ([CommonIapPurchase]?, (any Error)?) -> Void) {
        Task {
            do {
                let result = try await __fetchPurchases()
                completionHandler(result, nil)
            } catch {
                completionHandler(nil, error)
            }
        }
    }

    @nonobjc
    public func __fetchPurchases() async throws -> [CommonIapPurchase] {
        return try await fetchPurchases()
    }

    public func __fetchServiceAvailability(completionHandler: @escaping (IapServiceAvailability?, (any Error)?) -> Void) {
        Task {
            do {
                let result = try await __fetchServiceAvailability()
                completionHandler(result, nil)
            } catch {
                completionHandler(nil, error)
            }
        }
    }

    @nonobjc
    public func __fetchServiceAvailability() async throws -> IapServiceAvailability {
        return try await fetchServiceAvailability()
    }

    public func __isAuthorizedUser(completionHandler: @escaping (KotlinBoolean?, (any Error)?) -> Void) {
        Task {
            do {
                let result = try await __isAuthorizedUser()
                completionHandler(result, nil)
            } catch {
                completionHandler(nil, error)
            }
        }
    }

    @nonobjc
    public func __isAuthorizedUser() async throws -> KotlinBoolean {
        return try await isAuthorizedUser()
    }

    public func __purchaseProduct(params: CommonIapProductPurchaseParams, completionHandler: @escaping ((any IapPaymentResult)?, (any Error)?) -> Void) {
        Task {
            do {
                let result = try await __purchaseProduct(params: params)
                completionHandler(result, nil)
            } catch {
                completionHandler(nil, error)
            }
        }
    }

    @nonobjc
    public func __purchaseProduct(params: CommonIapProductPurchaseParams) async throws -> any IapPaymentResult {
        return try await purchaseProduct(params: params)
    }

    public func confirmPurchase(purchaseId: String, developerPayload: String?, completionHandler: @escaping (NSError?) -> Void) {
        Task {
            do {
                try await confirmPurchase(purchaseId: purchaseId, developerPayload: developerPayload)
                completionHandler(nil)
            } catch {
                completionHandler(error as NSError)
            }
        }
    }

    public func confirmPurchase(purchaseId: String, developerPayload: String?) async throws {
        // Your real implementation here
    }

    public func deletePurchase(purchaseId: String, completionHandler: @escaping (NSError?) -> Void) {
        Task {
            do {
                try await deletePurchase(purchaseId: purchaseId)
                completionHandler(nil)
            } catch {
                completionHandler(error as NSError)
            }
        }
    }

    public func deletePurchase(purchaseId: String) async throws {
        // Your real implementation here
    }

    public func fetchProducts(ids: [String], completionHandler: @escaping ([CommonIapProduct]?, NSError?) -> Void) {
        Task {
            do {
                let products = try await fetchProducts(ids: ids)
                completionHandler(products, nil)
            } catch {
                completionHandler(nil, error as NSError)
            }
        }
    }

    public func fetchProducts(ids: [String]) async throws -> [CommonIapProduct] {
        return []
    }

    public func fetchPurchaseInfo(purchaseId: String, completionHandler: @escaping (CommonIapPurchase?, NSError?) -> Void) {
        Task {
            do {
                let purchase = try await fetchPurchaseInfo(purchaseId: purchaseId)
                completionHandler(purchase, nil)
            } catch {
                completionHandler(nil, error as NSError)
            }
        }
    }

    public func fetchPurchaseInfo(purchaseId: String) async throws -> CommonIapPurchase? {
        return nil
    }

    public func fetchPurchases(completionHandler: @escaping ([CommonIapPurchase]?, NSError?) -> Void) {
        Task {
            do {
                let purchases = try await fetchPurchases()
                completionHandler(purchases, nil)
            } catch {
                completionHandler(nil, error as NSError)
            }
        }
    }

    public func fetchPurchases() async throws -> [CommonIapPurchase] {
        return []
    }

    public func fetchServiceAvailability(completionHandler: @escaping (IapServiceAvailability?, NSError?) -> Void) {
        Task {
            do {
                let availability = try await fetchServiceAvailability()
                completionHandler(availability, nil)
            } catch {
                completionHandler(nil, error as NSError)
            }
        }
    }

    public func fetchServiceAvailability() async throws -> IapServiceAvailability {
        return IapServiceAvailabilityUnavailable(throwable: nil)
    }

    public func fetchStore() -> CommonStore {
        return CommonStore.none
    }

    public func doInit(activity: CommonPlatformActivity) {
        // Implement if needed
    }

    public func isAuthorizedUser(completionHandler: @escaping (KotlinBoolean?, NSError?) -> Void) {
        Task {
            do {
                let result = try await isAuthorizedUser()
                completionHandler(result, nil)
            } catch {
                completionHandler(nil, error as NSError)
            }
        }
    }

    public func isAuthorizedUser() async throws -> KotlinBoolean {
        return KotlinBoolean(bool: false)
    }

    public func proceedIntent(intent: CommonPlatformIntent?, requestCode: KotlinInt?) {
        // Implement if needed
    }

    public func purchaseProduct(params: CommonIapProductPurchaseParams, completionHandler: @escaping (IapPaymentResult?, NSError?) -> Void) {
        Task {
            do {
                let result = try await purchaseProduct(params: params)
                completionHandler(result, nil)
            } catch {
                completionHandler(nil, error as NSError)
            }
        }
    }

    public func purchaseProduct(params: CommonIapProductPurchaseParams) async throws -> IapPaymentResult {
        return IapPaymentResultInvalidPaymentState()
    }
}

File: androidApp/src/rustore/java/ru/aleshin/studyassistant/data/IapServiceImpl.kt
/*
 * Copyright 2024 Stanislav Aleshin
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ru.aleshin.studyassistant.data

import android.app.Activity
import android.content.Context
import ru.aleshin.studyassistant.core.common.platform.PlatformActivity
import ru.aleshin.studyassistant.core.common.platform.services.iap.IapFailure
import ru.aleshin.studyassistant.core.common.platform.services.iap.IapPaymentResult
import ru.aleshin.studyassistant.core.common.platform.services.iap.IapProduct
import ru.aleshin.studyassistant.core.common.platform.services.iap.IapProductPurchaseParams
import ru.aleshin.studyassistant.core.common.platform.services.iap.IapPurchase
import ru.aleshin.studyassistant.core.common.platform.services.iap.IapService
import ru.aleshin.studyassistant.core.common.platform.services.iap.IapServiceAvailability
import ru.aleshin.studyassistant.core.common.platform.services.iap.IapServiceError
import ru.aleshin.studyassistant.core.common.platform.services.iap.PlatformIntent
import ru.aleshin.studyassistant.core.common.platform.services.iap.Store
import ru.aleshin.studyassistant.data.mappers.convertToCommon
import ru.rustore.sdk.billingclient.RuStoreBillingClient
import ru.rustore.sdk.billingclient.model.purchase.PurchaseAvailabilityResult
import ru.rustore.sdk.billingclient.usecase.ProductsUseCase
import ru.rustore.sdk.billingclient.usecase.PurchasesUseCase
import ru.rustore.sdk.billingclient.usecase.UserInfoUseCase
import ru.rustore.sdk.core.exception.RuStoreApplicationBannedException
import ru.rustore.sdk.core.exception.RuStoreException
import ru.rustore.sdk.core.exception.RuStoreNotInstalledException
import ru.rustore.sdk.core.exception.RuStoreOutdatedException
import ru.rustore.sdk.core.exception.RuStoreUserBannedException
import ru.rustore.sdk.core.exception.RuStoreUserUnauthorizedException
import ru.rustore.sdk.core.tasks.Task
import java.lang.ref.WeakReference

/**
 * @author Stanislav Aleshin on 16.06.2025.
 */
class IapServiceImpl(
    private val applicationContext: Context,
    private val billingClient: RuStoreBillingClient,
    private val purchaseInteractor: PurchasesUseCase = billingClient.purchases,
    private val productInteractor: ProductsUseCase = billingClient.products,
    private val userInteractor: UserInfoUseCase = billingClient.userInfo,
) : IapService {

    private var activity: WeakReference<Activity?> = WeakReference(null)

    private companion object {
        const val PACKAGE_NAME = "ru.aleshin.studyassistant"
        const val CONFIRM_SUB_BASE_URL = "https://public-api.rustore.ru/public/glike/subscription"
        const val JWE_TOKEN_HEADER = "Public-Token"
    }

    override fun init(activity: PlatformActivity) {
        this.activity = WeakReference(activity)
    }

    override fun fetchStore(): Store {
        return Store.RU_STORE
    }

    override suspend fun isAuthorizedUser(): Boolean {
        return userInteractor.getAuthorizationStatus().await().authorized
    }

    override suspend fun fetchServiceAvailability(): IapServiceAvailability {
        val result = purchaseInteractor.checkPurchasesAvailability().await()
        return when (result) {
            is PurchaseAvailabilityResult.Available -> {
                IapServiceAvailability.Available
            }
            is PurchaseAvailabilityResult.Unavailable -> {
                IapServiceAvailability.Unavailable(result.cause)
            }
            is PurchaseAvailabilityResult.Unknown -> {
                IapServiceAvailability.Unavailable(null)
            }
        }
    }

    override fun proceedIntent(intent: PlatformIntent?, requestCode: Int?) {
        if (requestCode == null) billingClient.onNewIntent(intent)
    }

    override suspend fun purchaseProduct(params: IapProductPurchaseParams): IapPaymentResult {
        val result = purchaseInteractor.purchaseProduct(
            productId = params.productId,
            orderId = params.orderId,
            quantity = params.quantity,
            developerPayload = params.developerPayload ?: params.appUserId,
        ).handledAwait()
        return result.convertToCommon()
    }

    override suspend fun fetchProducts(ids: List<String>): List<IapProduct> {
        return productInteractor.getProducts(ids).handledAwait().map { it.convertToCommon() }
    }

    override suspend fun fetchPurchaseInfo(purchaseId: String): IapPurchase? {
        return purchaseInteractor.getPurchaseInfo(purchaseId).handledAwait().convertToCommon()
    }

    override suspend fun fetchPurchases(): List<IapPurchase> {
        return purchaseInteractor.getPurchases().handledAwait().map { it.convertToCommon() }
    }

    override suspend fun confirmPurchase(purchaseId: String, developerPayload: String?) {
        purchaseInteractor.confirmPurchase(purchaseId, developerPayload).handledAwait()
    }

    override suspend fun deletePurchase(purchaseId: String) {
        purchaseInteractor.deletePurchase(purchaseId).handledAwait()
    }

    private fun <T> Task<T>.handledAwait(): T {
        return try {
            await()
        } catch (exception: Exception) {
            exception.printStackTrace()
            val type = when (exception) {
                is RuStoreNotInstalledException -> IapFailure.RuStoreNotInstalled
                is RuStoreOutdatedException -> IapFailure.RuStoreOutdated
                is RuStoreUserUnauthorizedException -> IapFailure.RuStoreUserUnauthorized
                is RuStoreApplicationBannedException -> IapFailure.RuStoreApplicationBanned
                is RuStoreUserBannedException -> IapFailure.RuStoreUserBanned
                is RuStoreException -> IapFailure.RuStoreBaseError
                else -> IapFailure.UnknownError
            }
            throw IapServiceError(type)
        }
    }
}